stablehlo/transforms/StablehloAggressiveFolder.cpp

```diff
@@ -643,6 +643,57 @@ struct EvalIotaOpPattern : public OpRewritePattern<IotaOp> {
   }
 };
 
+template <typename RangeType>
+DenseElementsAttr transposeType(TransposeOp& op, const RangeType& data) {
+  using ElementType = std::decay_t<decltype(*std::begin(data))>;
+
+  RankedTensorType operandType = op.getOperand().getType();
+  RankedTensorType resultType = op.getResult().getType();
+
+  const auto operandStrides = computeStrides(operandType.getShape());
+  const auto resultStrides = computeStrides(resultType.getShape());
+  const auto inversePermutation = invertPermutationVector(op.getPermutation());
+
+  SmallVector<ElementType> result;
+  result.reserve(resultType.getNumElements());
+
+  for (int64_t i = 0; i < resultType.getNumElements(); ++i) {
```

Comment:
Can we check for identity permutation 
```
bool isIdentityPermutation(const std::vector<unsigned>& permutation) {
   std::vector<unsigned> expected(permutation.size());
   std::iota(expected.begin(), expected.end(), 0); // Fill with 0, 1, 2, ...
   return permutation == expected;
  }

  return true;
}
```

in which case 
```
if (isIdentityPermutation(op.getPermutation())) {
    return DenseElementsAttr::get(resultType, data); 
  }
 ```

Reply:
That’s a good point! We can also check for splat values. However, I’m not sure if these checks should be placed in the folder; perhaps `StablehloAggressiveSimplification.cpp` would be a better location. To elaborate -- these fast paths could save some CPU time in the current implementation but in the end result would be the same. By placing them in `StablehloAggressiveSimplification.cpp`, users could simplify operations without the need to fold everything?

1) transpose(X) with an identity permutation => X
2) transpose(splat) with any permutation => splat

What do you think?

Reply:
The checks (including the splat one you mentioned) are equally valuable for opt-in passes like "aggressive-simplification" and folders invoked during operation creation using "OpBuilder::createOrFold," canonicalization, dialect conversion, and pattern rewriting with explicit calls to fold APIs. Unless we have a contract of "running the pass before folding," making both the pass and folder independently capable makes sense.

To achieve this, https://github.com/openxla/stablehlo/blob/main/stablehlo/transforms/PassUtils.cpp seems like a suitable place to hoist any common utilities, which in this case are the two canonicalization patterns you mentioned. The pass would utilize these canonicalization, and the folder could fold after applying them.
cc @GleasonK

Reply:
Let's leave these separate for now. Regardless of where they're implemented I think they should be separate patterns to keep things more maintainable / readable, and as such can be a separate PR. I.e. I'd rather we have `EvalTransposeSplat` and `EvalIdentityTranspose` than just an `EvalTranspose` method which does 10 things.

As for what constitutes folding vs canonicalizing..need to think more. I'd think a "simplification of a constant" should be a fold. Whereas `transpose(transpose(X))` should be a simplification. By that logic, `transpose(splat)` would be fold, and `transpose(X) ID -> X` is a simplification.